packages/angular_devkit/benchmark/src/run-benchmark-watch.ts

```diff
@@ -0,0 +1,115 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import { BaseException, logging } from '@angular-devkit/core';
+import { spawnSync } from 'child_process';
+import { Observable, forkJoin, of, throwError } from 'rxjs';
+import {
+  concatMap,
+  defaultIfEmpty,
+  first,
+  reduce,
+  repeat,
+  retryWhen,
+  startWith,
+  take,
+  takeUntil,
+  tap,
+  throwIfEmpty,
+  timeout,
+} from 'rxjs/operators';
+import { Command } from './command';
+import { MetricGroup } from './interfaces';
+import { LocalMonitoredProcess } from './monitored-process';
+import { MaximumRetriesExceeded, RunBenchmarkOptions } from './run-benchmark';
+import { aggregateMetricGroups } from './utils';
+
+export interface RunBenchmarkWatchOptions extends RunBenchmarkOptions {
+  watchMatcher: string;
+  watchTimeout?: number;
+  watchCommand: Command;
+}
+
+export function runBenchmarkWatch({
+  command, captures, reporters = [], iterations = 5, retries = 5, logger = new logging.NullLogger(),
+  watchMatcher, watchTimeout = 10000, watchCommand,
+}: RunBenchmarkWatchOptions): Observable<MetricGroup[]> {
+
+  let successfulRuns = 0;
+  let failedRuns = 0;
+  const debugPrefix = () => `Run #${successfulRuns + 1}:`;
+
+  // Run the process and captures, wait for both to finish, and average out the metrics.
+  const monitoredProcess = new LocalMonitoredProcess(command, false);
+  const processFailed = new BaseException('Wrong exit code.');
+
+  const stats$ = monitoredProcess.stats$.pipe(
+    takeUntil(monitoredProcess.stdout$.pipe(
+      first(stdout => stdout.toString().includes(watchMatcher)),
+      timeout(watchTimeout),
+    )),
+  );
+
+  return monitoredProcess.run()
```

Comment:
I feel a bit confused about the logic in this file, I guess because I haven't looked at this package in a while. 

Could you add some comments for readers here when you feel the PR isn't in draft anymore?

Reply:
Done